rasa/core/story_conflict.py

```diff
@@ -0,0 +1,252 @@
+        trackers: List["TrackerWithCachedStates"],
+        domain: Domain,
+        max_history: int,
+        rules: Dict[int, Optional[List[Text]]],
+    ):
+        # Iterate once more over all states and note the (unhashed) state,
+        # for which a conflict occurs
+        conflicts = {}
+        for tracker, event, sliced_states in StoryConflict._sliced_states_iterator(
+            trackers, domain, max_history
+        ):
+            h = hash(str(list(sliced_states)))
+
+            if h in rules and h not in conflicts:
+                conflicts[h] = StoryConflict(sliced_states)
+
+            if h in rules:
+                conflicts[h].add_conflicting_action(
+                    action=event.as_story_string(), story_name=tracker.sender_id
+                )
+
+        # Remove conflicts that arise from unpredictable actions
+        return [c for (h, c) in conflicts.items() if c.has_prior_events]
+
+    @staticmethod
+    def _sliced_states_iterator(
+        trackers: List[TrackerWithCachedStates], domain: Domain, max_history: int
+    ) -> (TrackerWithCachedStates, Event, List[Dict[Text, float]]):
+        """
+        Iterate over all given trackers and all sliced states within
+        each tracker, where the slicing is based on `max_history`
+        :param trackers: List of trackers
+        :param domain: Domain (used for tracker.past_states)
+        :param max_history: Assumed `max_history` value for slicing
+        :return: Yields (tracker, event, sliced_states) triplet
+        """
+        for tracker in trackers:
+            states = tracker.past_states(domain)
+            states = [
+                dict(state) for state in states
+            ]  # From rasa/core/featurizers.py:318
+
+            idx = 0
+            for event in tracker.events:
+                if isinstance(event, ActionExecuted):
+                    sliced_states = MaxHistoryTrackerFeaturizer.slice_state_history(
+                        states[: idx + 1], max_history
+                    )
+                    yield tracker, event, sliced_states
+                    idx += 1
+
+    @staticmethod
+    def _get_prev_event(
+        state: Optional[Dict[Text, float]]
+    ) -> [Optional[Text], Optional[Text]]:
+        """
+        Returns the type and name of the event (action or intent) previous to the
+        given state
+        :param state: Element of sliced states
+        :return: (type, name) strings of the prior event
+        """
+        prev_event_type = None
+        prev_event_name = None
+
+        if not state:
+            return prev_event_type, prev_event_name
+
+        for k in state:
+            if (
+                k.startswith(PREV_PREFIX)
+                and k[len(PREV_PREFIX) :] != ACTION_LISTEN_NAME
+            ):
+                prev_event_type = "action"
+                prev_event_name = k[len(PREV_PREFIX) :]
+
+            if not prev_event_type and k.startswith(MESSAGE_INTENT_ATTRIBUTE + "_"):
```

Comment:
I think there must be a slicker way to do this, e.g. even
`k.replace(MESSAGE_INTENT_ATTRIBUTE + "_", "")`  would make it easier to read